### PR TITLE
Rust: parse bytes like the C++ version and use `impl Trait` syntax

### PR DESCRIPTION
**Pull request by ricvelozo**
The C++'s `std::string` is encoding-agnostic.

### CODE REVIEW

File: brainfuck/bf.rs [hunk 1]
```diff
@@ -136,9 +136,9 @@ struct Program {
 }
 
 impl Program {
-    fn new(code: &str) -> Self {
+    fn new(code: &[u8]) -> Self {
         Self {
-            ops: parse(&mut code.chars()),
+            ops: parse(&mut code.iter().copied()),
```
Comment (nuald) [inline brainfuck/bf.rs:141]: Can we avoid copying by owning the `code`? It's not used anywhere else. Require more refactoring though, so it's minor.

File: brainfuck/bf.rs [hunk 2]
```diff
@@ -90,7 +90,7 @@ impl<'a> Printer<'a> {
         }
     }
 
-    fn get_checksum(&self) -> i32 {
+    const fn get_checksum(&self) -> i32 {
```
Comment (nuald) [inline brainfuck/bf.rs:93]: I don't see that function used in the const context, am I missing something here?

File: brainfuck/bf.rs (hunk 1 above)
Comment (ricvelozo) [reply brainfuck/bf.rs:141]: I tried to use `Vec`/`Iterator` and it became slower by 1 sec. Copying here is cheaper.

File: brainfuck/bf.rs (hunk 2 above)
Comment (ricvelozo) [reply brainfuck/bf.rs:93]: Yeah, currently nowhere can benefit of it.

File: brainfuck/bf.rs (hunk 1 above)
Comment (ricvelozo) [reply brainfuck/bf.rs:141]: Passing by reference, like:

```rust
fn parse<'a>(it: &mut impl Iterator<Item = &'a u8>) -> Box<[Op]> {
    // ...
}
```

Made no difference in speed (`mandel.b` runs in ~23.4 sec in my machine).

**Pull request merged**